frame/mosaic-vault/src/lib.rs

```diff
@@ -0,0 +1,904 @@
+
+		UnsupportedToken,
+
+		Underflow,
+
+		Overflow,
+
+		UnlockAmountGreaterThanTotalValueTransferred,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+
+		#[pallet::weight(10_000)]
+		 pub fn add_supported_token(origin: OriginFor<T>, 
+			asset_id: T::AssetId, 
+			remote_asset_id: T::RemoteAssetId, 
+			remote_network_id: T::RemoteNetworkId, 
+			max_asset_transfer_size: T::Balance,
+			min_asset_transfer_size: T::Balance,) -> DispatchResultWithPostInfo {
+           
+		   T::AdminOrigin::ensure_origin(origin)?;
+
+		   ensure!(max_asset_transfer_size > min_asset_transfer_size, Error::<T>::MaxAssetTransferSizeBelowMinimum);
+
+		   <RemoteAssetId<T>>::insert(remote_network_id, asset_id, remote_asset_id);	
+		   
+		   <MaxAssetTransferSize<T>>::insert(asset_id, max_asset_transfer_size);
+
+		   <MinAssetTransferSize<T>>::insert(asset_id, min_asset_transfer_size);
+
+		   Self::deposit_event(Event::TokenAdded{asset_id, remote_asset_id, remote_network_id});
+
+		   Ok(().into())
+		 }
+
+		 #[pallet::weight(10_000)]
+		 pub fn remove_supported_token(origin: OriginFor<T>, asset_id: T::AssetId, remote_network_id: T::RemoteNetworkId) -> DispatchResultWithPostInfo {
+
+			T::AdminOrigin::ensure_origin(origin)?; 
+
+			Self::only_supported_remote_token(remote_network_id.clone(), asset_id.clone())?;
+          
+ 		    let remote_asset_id = RemoteAssetId::<T>::get(remote_network_id, asset_id);
+
+		     <RemoteAssetId<T>>::remove(remote_network_id, asset_id);
+
+			 <MaxAssetTransferSize<T>>::remove(asset_id);
+
+			 <MinAssetTransferSize<T>>::remove(asset_id);
+
+             Self::deposit_event(Event::TokenRemoved{asset_id, remote_asset_id,  remote_network_id});
+
+			 Ok(().into())
+		 }
+
+		 #[pallet::weight(10_000)]
+		 pub fn set_asset_max_transfer_size(origin: OriginFor<T>, asset_id: T::AssetId, size: T::Balance) -> DispatchResultWithPostInfo {
+
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			 <MaxAssetTransferSize<T>>::insert(asset_id, size);
+
+			 Self::deposit_event(Event::AssetMaxTransferSizeChanged {asset_id, size});
+
+			 Ok(().into())
+		 }
+
+		 #[pallet::weight(10_000)]
+		 pub fn set_asset_min_transfer_size(origin: OriginFor<T>, asset_id: T::AssetId, size: T::Balance) -> DispatchResultWithPostInfo {
+
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			 <MinAssetTransferSize<T>>::insert(asset_id, size);
+
+			 Self::deposit_event(Event::AssetMinTransferSizeChanged {asset_id, size});
+
+			 Ok(().into())
+		 }
+
+		 #[pallet::weight(10_000)]
+		 pub fn set_transfer_lockup_time(origin: OriginFor<T>, lockup_time: Timestamp) -> DispatchResultWithPostInfo {
+
+			T::AdminOrigin::ensure_origin(origin.clone())?;
+			
+			let sender = ensure_signed(origin)?;
+
+			 let old_lockup_time = <TransferLockupTime<T>>::get();
+
+			 <TransferLockupTime<T>>::put(lockup_time);
+
+			 let action = "Transfer".as_bytes().to_vec();
+
+			 Self::deposit_event(Event::LockupTimeChanged{sender, old_lockup_time, lockup_time, action});
```

Comment:
not sure, why would you pass this string? I would prever `TransferLockupTimeChanged` event

Reply:
just to keep the events same as the solidity code, but what you said makes sense , ill check with them if its okay to omit the string